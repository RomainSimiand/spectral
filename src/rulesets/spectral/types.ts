export interface ISchema {
  [key: string]: IValue | IEnum | IOr | IObject | IArray;
}

export interface IValue {
  type: 'object' | 'enum' | 'array' | 'string' | 'boolean' | 'or' | 'any';
  optional?: boolean;
}

export interface IEnum extends IValue {
  type: 'enum';
  values?: any[];
}

export interface IOr extends IValue {
  type: 'or';
  types?: IValue[];
}

export interface IObject extends IValue {
  type: 'object';
  value?: ISchema;
}

export interface IArray extends IValue {
  type: 'array';
  of?: IValue[];
}
