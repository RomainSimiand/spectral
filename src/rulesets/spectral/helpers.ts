import { IArray, IObject, ISchema } from './types';

export function optional(type: any) {
  return {
    ...type,
    optional: true,
  };
}

export function object(value: ISchema): IObject {
  return {
    type: 'object',
    value,
  };
}

export function anEnum(values: any[]) {
  return {
    type: 'enum',
    values,
  };
}

export function primitive(type: 'string' | 'boolean' | 'any') {
  return {
    type,
  };
}

export function array(of: any): IArray {
  return {
    type: 'array',
    of,
  };
}
