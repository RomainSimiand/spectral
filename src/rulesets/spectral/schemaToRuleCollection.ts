import { DiagnosticSeverity } from '@stoplight/types';
import { RuleCollection, RuleFunction, RuleType } from '../../types';
import { ISchema } from './types';

export function noUndefRule(field: string) {
  return `${field}:no-undef`;
}

export function schemaToRuleCollection(schema: ISchema): RuleCollection {
  const ruleset: RuleCollection = {};
  Object.entries(schema).forEach(([key, value]) => {
    if (!value.optional) {
      ruleset[noUndefRule(key)] = {
        severity: DiagnosticSeverity.Error,
        type: RuleType.VALIDATION,
        summary: `'${key}' must be defined`,
        given: '$.rules[*]',
        then: {
          field: key,
          function: RuleFunction.TRUTHY,
        },
      };
    }
    switch (value.type) {
      case 'enum':
        break;
    }
  });
  return ruleset;
}
