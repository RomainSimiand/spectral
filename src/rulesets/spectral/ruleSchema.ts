import { anEnum, array, object, optional, primitive } from './helpers';
import { ISchema } from './types';

export const ruleSchema: ISchema = {
  type: optional(anEnum(['validation', 'style'])),
  summary: optional(primitive('string')),
  description: optional(primitive('string')),
  severity: optional(anEnum([0, 1, 2, 3])),
  tags: optional(array(primitive('string'))),
  enabled: optional(primitive('boolean')),
  given: primitive('string'),
  when: optional(
    object({
      field: primitive('string'),
      pattern: optional(primitive('string')),
    })
  ),
  then: {
    type: 'or',
    types: [
      object({
        field: optional(primitive('string')),
        function: primitive('string'),
        functionOptions: optional(primitive('any')),
      }),
      array(
        object({
          field: optional(primitive('string')),
          function: primitive('string'),
          functionOptions: optional(primitive('any')),
        })
      ),
    ],
  },
};
