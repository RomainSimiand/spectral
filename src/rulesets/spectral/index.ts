import { RuleCollection } from '../../types';
import { ruleSchema } from './ruleSchema';
import { schemaToRuleCollection } from './schemaToRuleCollection';

export const spectralRules = (): RuleCollection => schemaToRuleCollection(ruleSchema);
