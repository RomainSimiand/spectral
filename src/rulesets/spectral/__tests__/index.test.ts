import { pick } from 'lodash';
import { spectralRules } from '..';
import { Spectral } from '../../../spectral';
import { noUndefRule } from '../schemaToRuleCollection';

describe('Spectral Rules', () => {
  const rules = spectralRules();
  let spectral: Spectral;

  ['given', 'then'].forEach(field => {
    test(`rule is missing '${field}'`, async () => {
      spectral = new Spectral();
      spectral.addRules(pick(rules, noUndefRule(field)));

      const results = await spectral.run({
        rules: [
          {
            [field]: 'value',
          },
          {
            // field is missing
          },
        ],
      });
      expect(results).toMatchSnapshot();
    });
  });
});
