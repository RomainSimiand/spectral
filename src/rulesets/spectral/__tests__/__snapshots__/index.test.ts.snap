// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spectral Rules rule is missing 'given' 1`] = `
Array [
  Object {
    "code": "given:no-undef",
    "message": "rules.1.given is not truthy",
    "path": Array [
      "rules",
      1,
      "given",
    ],
    "range": Object {
      "end": Object {
        "character": 6,
        "line": 5,
      },
      "start": Object {
        "character": 4,
        "line": 5,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "'given' must be defined",
  },
]
`;

exports[`Spectral Rules rule is missing 'then' 1`] = `
Array [
  Object {
    "code": "then:no-undef",
    "message": "rules.1.then is not truthy",
    "path": Array [
      "rules",
      1,
      "then",
    ],
    "range": Object {
      "end": Object {
        "character": 6,
        "line": 5,
      },
      "start": Object {
        "character": 4,
        "line": 5,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "'then' must be defined",
  },
]
`;
